### PR TITLE
refactor WatchMojo and WatchMojoTest classes

### LINKED ISSUE #2638: WatchMojo : Remove child class fields shadowing parent class fields
### Component

Kubernetes Maven Plugin

### Description

We have a mojo class WatchMojo in kubernetes-maven-plugin that extends AbstractDockerMojo
https://github.com/eclipse/jkube/blob/62080bfd0ce8c65245fef69723c95816d49661bb/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojo.java#L67

It contains these two fields : 
https://github.com/eclipse/jkube/blob/62080bfd0ce8c65245fef69723c95816d49661bb/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojo.java#L75-L78

However, the parent class which it extends from also contains these fields, so WatchMojo should be already inheriting fields from the parent class
https://github.com/eclipse/jkube/blob/62080bfd0ce8c65245fef69723c95816d49661bb/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java#L346-L347

https://github.com/eclipse/jkube/blob/62080bfd0ce8c65245fef69723c95816d49661bb/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java#L291-L292

We need to remove child fields from child class that are same as parent class. Note that only `protected` access level fields are accessible to child class. Read [Sonar rule S2387](https://github.com/joansmith/sonar-java/blob/master/java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S2387.html) for more details

### Expected Behavior
Fields with colliding names are removed from WatchMojo

### Acceptance Criteria
- [ ] Fields with matching name in AbstractDockerMojo are removed from WatchMojo

### Before you start :red_circle:

:point_down: :point_down: :point_down: :point_down: :point_down: :point_down: :point_down: :point_down: :point_down: :point_down::point_down::point_down::point_down::point_down::point_down::point_down:
Make sure you read the [contributing guide](https://www.eclipse.org/jkube/contributing) first. Pay special attention to the ECA agreement section and the requirement to sign-off your commit.

#### Kubernetes

If you don't have a real Kubernetes cluster available (most probably), you can use Minikube or Kind to test with a local cluster.

- Minikube [get started guide](https://minikube.sigs.k8s.io/docs/start/)
- Kind (Requires Docker) [quick start](https://kind.sigs.k8s.io/docs/user/quick-start/)

#### OpenShift

If you don't have a real OpenShift cluster available (most probably), you can use [Red Hat's developer Sandbox for Red Hat OpenShift](https://developers.redhat.com/developer-sandbox). The only requirement is to have a Red Hat account.

Once you have your Sandbox environment, you'll need to download the oc tool from the cluster console.
(Press the `?` icon and from the context menu select `Command line tools`, you'll be redirected to https://$subdomain.openshiftapps.com/command-lines-tools where you'll be able to download the CLI for your platform)

### PR DESCRIPTION
**Pull request by arsenalzp**
## Description

Fixes #2638 

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [X] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.dev/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] My code follows the style guidelines of this project
 - [X] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [ ] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [ ] I have performed a self-review of my code
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [X] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] I have added tests that prove my fix is effective or that my feature works
 - [X] New and existing unit tests pass locally with my changes
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
### Eclipse JKube [CI Report](https://github.com/jkubeio/ci)

Started new GH workflow run for https://github.com/eclipse/jkube/pull/2729 (_2024-02-24T11:59:17Z_)

:gear: [JKube E2E Tests (8030344944)](https://github.com/jkubeio/ci/actions/runs/8030344944) :hourglass_flowing_sand:
- :hourglass_flowing_sand: [Bootstrap CI test run (#2729)](https://github.com/jkubeio/ci/actions/runs/8030344944/job/21937622439)

**Comment by arsenalzp** [comment]
Not sure why it failed, I've just ran test locally, here are results:

> [INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 213, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-failsafe-plugin:3.1.2:verify (default) @ gradle-plugin-it ---
[INFO] 
[INFO] --- maven-install-plugin:2.4:install (default-install) @ gradle-plugin-it ---
[INFO] Installing /home/arsenal/eclipse-workspace/jkube/gradle-plugin/it/target/gradle-plugin-it-1.17-SNAPSHOT.jar to /home/arsenal/.m2/repository/org/eclipse/jkube/gradle-plugin-it/1.17-SNAPSHOT/gradle-plugin-it-1.17-SNAPSHOT.jar
[INFO] Installing /home/arsenal/eclipse-workspace/jkube/gradle-plugin/it/pom.xml to /home/arsenal/.m2/repository/org/eclipse/jkube/gradle-plugin-it/1.17-SNAPSHOT/gradle-plugin-it-1.17-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Eclipse JKube 1.17-SNAPSHOT:
[INFO] 
[INFO] Eclipse JKube ...................................... SUCCESS [  0.459 s]
[INFO] JKube Kit :: Parent ................................ SUCCESS [  0.633 s]
[INFO] JKube Kit :: Documentation ......................... SUCCESS [ 10.269 s]
[INFO] JKube Kit :: API ................................... SUCCESS [  0.481 s]
[INFO] JKube Kit :: Common ................................ SUCCESS [ 29.583 s]
[INFO] JKube Kit :: Common Maven .......................... SUCCESS [  4.431 s]
[INFO] JKube Kit :: Config :: Image ....................... SUCCESS [  5.553 s]
[INFO] JKube Kit :: Config :: Resource .................... SUCCESS [  7.133 s]
[INFO] JKube Kit :: Build :: API .......................... SUCCESS [  5.919 s]
[INFO] JKube Kit :: Build :: Service :: Docker ............ SUCCESS [ 24.877 s]
[INFO] JKube Kit :: Build :: Service :: JIB ............... SUCCESS [  3.867 s]
[INFO] JKube Kit :: Build :: Service :: Buildpacks ........ SUCCESS [  4.948 s]
[INFO] JKube Kit :: Enricher :: API ....................... SUCCESS [ 10.783 s]
[INFO] JKube Kit :: Helm .................................. SUCCESS [ 18.060 s]
[INFO] JKube Kit :: Config :: Service ..................... SUCCESS [01:12 min]
[INFO] JKube Kit :: Generator :: API ...................... SUCCESS [  5.259 s]
[INFO] JKube Kit :: Generator :: Java Exec ................ SUCCESS [  4.966 s]
[INFO] JKube Kit :: Generator :: Karaf .................... SUCCESS [  2.749 s]
[INFO] JKube Kit :: Generator :: WebApp ................... SUCCESS [  4.801 s]
[INFO] JKube Kit :: Enricher :: Generic ................... SUCCESS [ 18.814 s]
[INFO] JKube Kit :: Enricher :: Specific .................. SUCCESS [  5.208 s]
[INFO] JKube Kit :: Profile ............................... SUCCESS [  2.158 s]
[INFO] JKube Kit :: Resource :: Service ................... SUCCESS [  5.426 s]
[INFO] JKube Kit :: Watcher :: API ........................ SUCCESS [  3.282 s]
[INFO] JKube Kit :: Watcher :: Standard ................... SUCCESS [  5.391 s]
[INFO] JKube Kit :: Remote Development .................... SUCCESS [ 16.713 s]
[INFO] JKube Kit :: Spring-Boot ........................... SUCCESS [ 12.856 s]
[INFO] JKube Kit :: Thorntail ............................. SUCCESS [  5.552 s]
[INFO] JKube Kit :: Vert.x ................................ SUCCESS [  7.821 s]
[INFO] JKube Kit :: Quarkus ............................... SUCCESS [  6.542 s]
[INFO] JKube Kit :: Micronaut ............................. SUCCESS [  4.299 s]
[INFO] JKube Kit :: Microprofile .......................... SUCCESS [  3.492 s]
[INFO] JKube Kit :: Open Liberty .......................... SUCCESS [  5.433 s]
[INFO] JKube Kit :: WildFly JAR ........................... SUCCESS [  4.201 s]
[INFO] JKube Kit :: SmallRye .............................. SUCCESS [  3.243 s]
[INFO] JKube Kit :: Helidon ............................... SUCCESS [  3.564 s]
[INFO] JKube Kit :: Common :: Integration Test Utils ...... SUCCESS [  1.703 s]
[INFO] JKube Kit .......................................... SUCCESS [  0.010 s]
[INFO] Kubernetes Maven Plugin :: Parent .................. SUCCESS [  0.005 s]
[INFO] Kubernetes Maven Plugin :: Plugin .................. SUCCESS [ 13.103 s]
[INFO] Kubernetes Maven Plugin :: Documentation ........... SUCCESS [  7.240 s]
[INFO] Kubernetes Maven Plugin :: Integration Tests ....... SUCCESS [04:40 min]
[INFO] OpenShift Maven Plugin:: Parent .................... SUCCESS [  0.064 s]
[INFO] OpenShift Maven Plugin :: Plugin ................... SUCCESS [  7.160 s]
[INFO] OpenShift Maven Plugin :: Integration Tests ........ SUCCESS [03:20 min]
[INFO] Gradle Plugin :: Parent ............................ SUCCESS [  0.079 s]
[INFO] Gradle Plugin :: Kubernetes ........................ SUCCESS [ 16.438 s]
[INFO] Gradle Plugin :: OpenShift ......................... SUCCESS [ 11.661 s]
[INFO] Gradle Plugin :: Documentation ..................... SUCCESS [  7.087 s]
[INFO] Gradle Plugin :: Integration Tests ................. SUCCESS [06:45 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  21:22 min
[INFO] Finished at: 2024-02-25T13:59:18+02:00
[INFO] ------------------------------------------------------------------------

File: kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java [hunk 1]
```diff
@@ -648,5 +649,20 @@ protected GavLabel getGavLabel() {
         // Label used for this run
         return new GavLabel(project.getGroupId(), project.getArtifactId(), project.getVersion());
     }
-
+    
+    public ResourceConfig getResources() {
+        return this.resources;
+    }
```
Comment (rohanKanojia) [inline kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java:655]: Umm, why do we need getters for these fields? These fields are `protected` so they should be  visible in classes extending this class.

Comment (arsenalzp) [reply kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java:655]: You are right it should be, however in my IDE I got error about visibility, thereforeI decided to use getters/setters methods. However let me run a single test without gettrs/setter and let's see :)

Comment (arsenalzp) [reply kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java:655]: Tried to run a test without getters, got compilation error:
> $ cd kubernetes-maven-plugin/plugin
$ mvn test -Dtest=WatchMojoTest
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/arsenal/eclipse-workspace/jkube/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojoTest.java:[123,14] resources has protected access in org.eclipse.jkube.maven.plugin.mojo.build.AbstractDockerMojo
[ERROR] /home/arsenal/eclipse-workspace/jkube/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojoTest.java:[135,14] namespace has protected access in org.eclipse.jkube.maven.plugin.mojo.build.AbstractDockerMojo
[INFO] 2 errors 
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  18.900 s
[INFO] Finished at: 2024-02-26T09:50:56+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:testCompile (default-testCompile) on project kubernetes-maven-plugin: Compilation failure: Compilation failure: 
[ERROR] /home/arsenal/eclipse-workspace/jkube/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojoTest.java:[123,14] resources has protected access in org.eclipse.jkube.maven.plugin.mojo.build.AbstractDockerMojo
[ERROR] /home/arsenal/eclipse-workspace/jkube/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojoTest.java:[135,14] namespace has protected access in org.eclipse.jkube.maven.plugin.mojo.build.AbstractDockerMojo

Comment (arsenalzp) [reply kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java]: Yes, sure! However what was the issue?

File: kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojoTest.java [hunk 2]
```diff
@@ -131,10 +126,27 @@ void executeInternal_whenNamespaceConfiguredInResourceConfig_shouldDelegateToWat
...
-    watchMojo.namespace = "configured-namespace";
+    testNamespace = "configured-namespace";
+    watchMojo = createWatchMojo();
     // When
     watchMojo.execute();
     // Then
     watcherManagerMockedStatic.verify(() -> WatcherManager.watch(any(), eq("configured-namespace"), any(), any()), times(1));
   }
+
+  private WatchMojo createWatchMojo() {
```
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojoTest.java:137]: okay, let me revert this to previous approach

File: kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java (hunk 1 above)
Comment (rohanKanojia) [reply kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java]: Sorry, I was trying a different approach of not relying on getters/setters at all, but as per https://github.com/eclipse/jkube/pull/2729#discussion_r1503844358 it makes the test slightly more difficult to maintain. I've reverted it to what you had done earlier.

I tried removing getters but I didn't get the abovementioned error.

**Comment by arsenalzp** [comment]
Can PR be merged? I would like to take another issue to work on, because I liked this project and community around it.

**Review by rohanKanojia — APPROVED** [review]

**Pull request merged**